src/library_pthread.js

```diff
@@ -1069,6 +1069,90 @@ var LibraryPThread = {
 #endif
   },
 
+#if MAIN_MODULE
+  $promiseMap: "new Map();",
+  $nextPromiseId: 1,
+
+  // Create a new promise that can be resolved or rejected by passing
+  // a unique ID to emscripten_promise_resolve/emscripten_promise_reject
+  // TODO(sbc): Should be factor this out and make this API public?
+  $newNativePromise__deps: ['$promiseMap', '$nextPromiseId'],
+  $newNativePromise: function(func, args) {
+    return new Promise((resolve, reject) => {
+      var promiseId = nextPromiseId;
+      nextPromiseId += 1;
```

Comment:
This is similar to what we do in the OPFS WasmFS backend, but there we use free lists (which I couldn't figure out how to share effectively) https://github.com/emscripten-core/emscripten/blob/main/src/library_wasmfs_opfs.js#L8-L71

Reply:
Thanks!